src/categorical_algebra/Automorphisms.jl

```diff
@@ -0,0 +1,306 @@
+                 for a in autos(p)], by=string)
+    return isempty(isos) ? g : isos[1]
+end
+
+function canonical_hash(g::ACSet)::UInt64
+    return hash(string(canonical_iso(g)))
+end
+
+# Store 1.) how many of each color (for each in arrow) has
+# a particular point as its target, 2.) what color (for
+# each out arrow) the data point is the src of
+CDataPoint = Tuple{Vector{Vector{Int}}, Vector{Int}}
+# Data req' to color a CSet (each element of each component)
+CData = Dict{Symbol, Vector{CDataPoint}}
+
+"""
+This could be extended to add extra automorphism-invariant
+properties in the output. Right now, we just compute the
+following pair for each element of each component:
+- number of incident edges from each arrow (from each color)
+- the color pointed to by each outgoing edge
+
+This does not generalize to ACSets. We cannot naively
+throw the attributes as raw data into the color data.
+It will make indistinguishable elements (e.g. two
+elements that map to different data but otherwise can
+be permuted) as distinguishable.
+"""
+function compute_color_data(g::CSet{CD}, color::CDict)::CData where {CD}
+    tabs, arrs, srcs, tgts = ob(CD), hom(CD), dom(CD), codom(CD)
+    ntab = eachindex(tabs)
+    res = CData()
+    a_in  = [[a_ind for (a_ind, t_ind) in enumerate(tgts) if t_ind==i] for i in ntab]
+    a_out = [[a_ind for (a_ind, s_ind) in enumerate(srcs) if s_ind==i] for i in ntab]
+
+    for (tgt_ind, tgt) in enumerate(tabs)
+        subres = Vector{Vector{Int}}[]  # for each table
+        for arr_ind in a_in[tgt_ind]
+            src = tabs[srcs[arr_ind]]
+            color_src = color[src]
+            n_color_src = max0(color_src)
+            preimg = g.indices[arrs[arr_ind]]
+            subsubres = Vector{Int}[]  # for particular elem in tgt
+            for tgt_elem in 1:nparts(g, tgt)
+                precolor_elem = color_src[preimg[tgt_elem]]
+                n_precolor = [count(==(x), precolor_elem) for x in 1:n_color_src]
+                push!(subsubres, n_precolor)
+            end
+            push!(subres, subsubres)
+        end
+
+        # Also compute per-element data for table `tgt` (now, think of it as a src)
+        outgoing_arrows = a_out[tgt_ind]
+        out_subres = Vector{Int}[color[tabs[tgts[oga]]][g[arrs[oga]]]
+                                 for oga in outgoing_arrows]
+
+        # Combine the two pieces of data for each elmeent in tgt, store in res
+        res[tgt] = [([ssr[i] for ssr in subres],
+                     [osr[i] for osr in out_subres])
+                    for i in 1:nparts(g,tgt)]
+    end
+    return res
+end
+
+"""
+Construct permutation σ⁻¹ such that σσ⁻¹=id
+"""
+function invert_perm(x::CDict)::CDict
+    function invert_comp(xs::Vector{Int})::Vector{Int}
+        return [findfirst(==(v), xs) for v in eachindex(xs)]
+    end
+    return Dict([k=>invert_comp(v) for (k, v) in collect(x)])
+end
+
+"""
+Compose permutations
+"""
+function compose_perms(x::CDict, y::CDict)::CDict
```

Comment:
This would be nice! I tried this - it requires some boxing/unboxing of vectors that made the rest of the code look more complicated (and introduced an infinite loop bug that I don't know how to fix), so I'm not sure this is worth it to save the line of duplicated code.

https://github.com/kris-brown/Catlab.jl/commit/c66ea2f1d36fff3462342359bb405df9deab1027